python/adelphi/setup.py

```diff
@@ -5,10 +5,10 @@
 PY3 = sys.version_info[0] == 3
 
 dependencies = [
-    'cassandra-driver ~= 3.24.0',
-    'click ~= 7.1.0',
+    'cassandra-driver ~= 3.24',
```

Comment:
I know this has come up multiple times before, but I'm a bit torn between being flexible with the driver version versus ensuring non-anonymized info won't leak.
What if a newer driver version adds more elements to the CQL output that we're not anonymizing? Or if it changes how it handles the object references internally?

I think the ideal approach would be to detach from the `cluster.metadata` object that we rely on straight out of the driver and instead deep clone `metadata` and copy over only the attributes we're anonymizing. That way we wouldn't have to worry about driver changes.

Reply:
I don't think the management of the cluster.metadata object is the problem in the scenario you describe @grighetto.  The fns in adelphi.anonymize swap names around in the various metadata objects returned by the driver.  Whether that's the actual cluster.metadata object itself or a clone of it doesn't really change much... you're still making some assumptions about the internal structures of the metadata object.  Well, I mean you kinda are... you're assuming certain fields exist and that they're writable, and while I agree that that does make you dependent upon the driver API I don't think you make the situation better by using a clone instead.

It seems like the way to accomplish what you're after would be to create some new structure (let's say just a map), populate it from metadata information and pass that along to anonymization.  If I read your comment right this is closer to what you were suggesting.  You could then insulate anonymization against changes in the driver API... but what good does that really do for you?  You've just pushed the problem up a step; now changes to the driver API break your extraction process but they don't break anonymization.  Is that... better?

To make matters worse all of this is overshadowed by the fact that we generate the CQL by calling the method export_as_string() on metadata objects.  Now we're no longer talking about the structure of metadata objects; we're now invoking logic.  That logic could do anything and could change between releases.  The changes described above provide exactly zero protection against that kind of change.  Neither, unfortunately, do our existing unit tests around anonymization.

So why do I not share your concern about these changes?  Two reasons really.

First, we also have integration tests around CQL generation which function as end-to-end tests.  If new material were added in newer driver versions any run against those new versions would indicate a mismatch.

Second, there is no active feature development on the Python driver (or any driver for that matter) at the moment because no team is working on them.  So it's extremely unlikely that any such feature would ever get scheduled let alone released.

So it seems like the concern here requires (a) work from a team that doesn't exist [1] to publish a new version that (b) breaks anonymization in some way and (c) somehow isn't subjected to our integration tests.  While I admit that such a scenario is logically possible the actual probability of all of that happening seems very, very small to me.

[1] This is a bit of an overstatement; a few minor patches are making it in and I think there's still one contractor working on some things.  But it's not much of an overstatement.

Reply:
> Whether that's the actual cluster.metadata object itself or a clone of it doesn't really change much

The "clone" would only contain the attributes we anonymize. If a newer driver version adds some information that we're not anonymizing, it wouldn't appear in the resulting CQL output.

> changes to the driver API break your extraction process but they don't break anonymization. Is that... better?

Definitely, the worst thing that could possibly happen is leaking information.

> That logic could do anything and could change between releases. The changes described above provide exactly zero protection against that kind of change.

As long as the clone contain only the information we want and we're not connected to the cluster, we should be safe with regards to leaking.

> we also have integration tests around CQL generation which function as end-to-end tests

How does that help adelphi installations out there that will pick up a newer driver version automatically?

> there is no active feature development on the Python driver

I wouldn't make any kind of technical decision based on that.

> While I admit that such a scenario is logically possible the actual probability of all of that happening seems very, very small to me.

Again, we must prevent any possibility of leaking info.